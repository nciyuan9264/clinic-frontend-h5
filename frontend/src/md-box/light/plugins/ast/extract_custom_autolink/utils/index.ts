export { HyperNodeType, getHyperLinkNodes } from './parser';

export const completeProtocol = (raw: string) => {
  if (raw.match(/https?:\/\//)) {
    return raw;
  }
  return `https://${raw}`;
};
