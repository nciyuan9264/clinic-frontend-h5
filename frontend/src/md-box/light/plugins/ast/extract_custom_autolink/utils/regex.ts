// -> refer @ea/url-recognition -> refer 飞书云文档
const PROTOCOL = 'https?|s?ftp|ftps|nfs|ssh';
const LINK_SUFFIX =
  "\\b([/#?][\\-A-Z0-9@:%_+.~#?&'/=;$,!\\*\\[\\]\\(\\){}^|<>]*[-A-Z0-9&@#/%=~\\(\\)_|]?)?";
const LOCALHOST_REG = /localhost(:[0-9]{2,5})?/;
const IP_ADDRESS_REG =
  // eslint-disable-next-line max-len
  /(([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\.){3}(2[0-4][0-9]|25[0-5]|1[0-9]{2}|[1-9][0-9]|[0-9])(:[0-9]{2,5})?/;
const IP_REG = new RegExp(
  `((((${PROTOCOL})://)?(${LOCALHOST_REG.source}))|(((${PROTOCOL})://)(${IP_ADDRESS_REG.source})))${LINK_SUFFIX}`,
  'gi',
);
const URL_HOST_BODY = '([-A-Z0-9%_+~#@][\\-A-Z:0-9%_+~#@]{0,256}\\.){1,50}';
const ANY_TOP_DOMAIN = '[a-z\\-]{2,15}';
const TOP_DOMAIN =
  'com|org|net|int|edu|gov|mil|ac|ad|ae|af|ag|ai|al|am|an|ao|aq|ar|as|at|au|aw|ax|az|ba|bb|bd|be|bf|bg|bh|bi|bj|bm|bn|bo|br|bs|bt|bv|bw|by|bz|ca|cc|cf|cg|ch|ci|ck|cl|cm|cn|co|cr|cu|cv|cw|cx|cy|cz|de|dj|dk|dm|do|dz|ec|ee|eg|eh|er|es|et|eu|fi|fj|fk|fm|fo|fr|ga|gb|gd|ge|gf|gh|gi|gl|gm|gn|gp|gr|gs|gt|gu|gw|gy|hk|hm|hn|hr|ht|hu|id|ie|il|im|in|io|iq|ir|is|it|jm|jo|jp|ke|kg|kh|ki|km|kn|kp|kr|kw|ky|kz|la|lb|lc|li|lk|lr|ls|lt|lu|lv|ly|ma|mc|md|me|mf|mg|mh|mk|ml|mm|mn|mo|mp|mq|mr|ms|mt|mu|mv|mw|mx|my|mz|na|nc|ne|nf|ng|ni|nl|no|np|nr|nu|nz|om|pa|pe|pf|pg|ph|pk|pl|pm|pn|pr|ps|pt|pw|py|qa|re|ro|rs|ru|rw|sa|sb|sc|sd|se|sg|sh|si|sj|sk|sl|sm|sn|so|sr|ss|st|su|sv|sx|sy|sz|tc|td|tf|tg|th|tj|tk|tl|tm|tn|to|tp|tr|tt|tv|tw|tz|ua|ug|uk|um|us|uy|uz|va|vc|ve|vg|vi|vn|vu|wf|ws|ye|yt|za|zm|zw|site|top|wtf|xxx|xyz|cloud|engineering|help|one';
const TOP_DOMAIN_LITE = 'com|cn|tk|de|net|org|uk|info|nl|ru|hk';
// 由于顶级域名不全，故更改为带 协议头的url 忽略上面顶级域名，直接匹配
const URL_REG = new RegExp(
  // eslint-disable-next-line max-len
  `(((${PROTOCOL}):\\/\\/${URL_HOST_BODY}${ANY_TOP_DOMAIN})|(${URL_HOST_BODY}(${TOP_DOMAIN_LITE})))(:[0-9]{2,5})?${LINK_SUFFIX}`,
  'gi',
);
// mailto的贪婪匹配在遇到极长的字符串（无空格的英文）会很慢，需要限长。
const EMAIL_URL_BODY = "[\\w.!#$%&'*+-/=?^_\\`{|}~]{1,2000}@[A-Za-z0-9_.-]+\\.";
const LENGTH_LIMITED_EMAIL_REG = new RegExp(
  `^(?!data:)((mailto:${EMAIL_URL_BODY}${ANY_TOP_DOMAIN})|(${EMAIL_URL_BODY}(${TOP_DOMAIN})))\\b`,
  'gi',
);

export const urlRegExp = new RegExp(
  `(${IP_REG.source})|(${URL_REG.source})|(${LENGTH_LIMITED_EMAIL_REG.source})`,
  'gi',
);
