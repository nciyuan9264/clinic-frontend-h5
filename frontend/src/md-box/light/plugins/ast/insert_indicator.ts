import type { HTML } from 'mdast';

import {
  AstPluginConfig,
  AstPluginModifier,
  BaseAstPlugin,
  isLiteralOfContent,
  isImage,
  isParentOfContent,
  stringifyChildren,
  AstPluginPriority,
} from '../../utils';

const indicatorItem: HTML = {
  type: 'html',
  value: '<span class="indicator" />',
};

export class InsertIndicatorAstPlugin extends BaseAstPlugin {
  name = 'insert_indicator';

  config: AstPluginConfig = {
    isReverse: true,
    priority: AstPluginPriority.After,
  };

  modifier: AstPluginModifier = (item, { insertAfter, stop }) => {
    /** 图片后面不插入指示点 */
    if (isImage(item)) {
      stop();
      return;
    }
    /** 空列表项则直接在列表项内插入指示点 */
    if (
      isParentOfContent(item) &&
      item.type === 'listItem' &&
      !stringifyChildren(item.children, true).trim()
    ) {
      stop();
      insertAfter(
        [
          {
            ...item,
            children: [indicatorItem],
          },
        ],
        true,
      );

      return;
    }
    /** 非code的叶子结点类型则后面插入指示点 */
    if (isLiteralOfContent(item)) {
      stop();
      if (item.type === 'code') {
        return;
      }
      insertAfter([indicatorItem]);
    }
  };
}
