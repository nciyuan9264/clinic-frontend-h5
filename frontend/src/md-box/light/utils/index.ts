export * from './micromark';
export * from './marked';
export * from './renderer';
export * from './common';
export * from './hop';
export * from './plugin';
