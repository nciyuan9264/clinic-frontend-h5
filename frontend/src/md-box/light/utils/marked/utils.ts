/** 源自 markedjs 源码 */

const escapeTest = /[&<>"']/;
const escapeReplace = new RegExp(escapeTest.source, 'g');
const escapeTestNoEncode = /[<>"']|&(?!(#\d{1,7}|#[Xx][a-fA-F0-9]{1,6}|\w+);)/;
const escapeReplaceNoEncode = new RegExp(escapeTestNoEncode.source, 'g');

const escapeReplacements: Record<string, string> = {
  '&': '&amp;',
  '<': '&lt;',
  '>': '&gt;',
  '"': '&quot;',
  "'": '&#39;',
};

export const escapeMarkdownText = (html: string, encode = false) => {
  const getEscapeReplacement = (ch: string) => escapeReplacements[ch];

  if (encode) {
    if (escapeTest.test(html)) {
      return html.replace(escapeReplace, getEscapeReplacement);
    }
  } else {
    if (escapeTestNoEncode.test(html)) {
      return html.replace(escapeReplaceNoEncode, getEscapeReplacement);
    }
  }
  return html;
};
