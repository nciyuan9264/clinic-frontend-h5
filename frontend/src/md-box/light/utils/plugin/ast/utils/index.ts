export * from './update_ast';
