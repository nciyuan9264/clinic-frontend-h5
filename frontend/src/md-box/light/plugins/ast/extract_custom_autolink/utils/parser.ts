import emailValidator from 'email-validator';

import { urlRegExp } from './regex';

export enum HyperNodeType {
  text,
  link,
}

export type HyperNode =
  | {
      type: HyperNodeType.text;
      text: string;
    }
  | {
      type: HyperNodeType.link;
      url: string;
    };

export type HyperNodeTitled = HyperNode & {
  title: string;
};

const openAndCloseCharPairs: Record<string, string> = {
  '(': ')',
  '[': ']',
  '{': '}',
  '<': '>',
};
const openChars = Object.keys(openAndCloseCharPairs);
const closeChars = Object.values(openAndCloseCharPairs);
const unreasonableSuffix = ['.', ',', '!', '?', "'", ':', ';', '@', '#'];

/**
 * 分割合理的URL
 * @returns 返回字符串中合理的URL的长度，范围[0, input.length]
 */
const getAReasonableUrl = (input: string) => {
  let index = -1;
  const stack: Array<{
    char: string;
    index: number;
  }> = [];
  for (let i = 0; i < input.length; ++i) {
    const char = input[i];
    if (openChars.includes(char)) {
      stack.push({
        char,
        index: i,
      });
    } else if (closeChars.includes(char)) {
      if (!stack.length) {
        index = i;
        break;
      } else {
        const { char: lastOpenChar } = stack[stack.length - 1];
        if (char === openAndCloseCharPairs[lastOpenChar]) {
          stack.pop();
        } else {
          ({ index } = stack[0]);
          break;
        }
      }
    }
  }
  if (stack.length) {
    ({ index } = stack[0]);
  }
  if (index === -1) {
    index = input.length;
  }

  while (index > 1 && unreasonableSuffix.includes(input[index - 1])) {
    index--;
  }
  return input.slice(0, index);
};

/**
 * 分割字符串中的超链接和常规字符，输出格式化数据 HyperNode[]
 */
export const getHyperLinkNodes = (str: string): HyperNode[] => {
  const nodes: HyperNode[] = [];
  let cur = 0;
  urlRegExp.lastIndex = 0;
  while (true) {
    const result = urlRegExp.exec(str);
    if (!result) {
      break;
    }

    const { index, 0: content } = result;
    const text = str.slice(cur, index);

    // ① 文字部分
    if (text) {
      nodes.push({
        type: HyperNodeType.text,
        text,
      });
    }

    // ② 链接/markdown/email 部分，email 被渲染为文字，code block / inline code 中的链接也当做文字
    if (content) {
      /** 1. 非正常前缀 */
      const abnormalPrefixMatchResult = /^(@|#|\+|-|%)+/.exec(content);
      if (abnormalPrefixMatchResult) {
        nodes.push({
          type: HyperNodeType.text,
          text: abnormalPrefixMatchResult[0],
        });
        cur = result.index + abnormalPrefixMatchResult[0].length;
        urlRegExp.lastIndex = cur;
      } else {
        /** 2. 合理Url作为前缀 */
        const subUrl = getAReasonableUrl(content);

        if (subUrl) {
          if (emailValidator.validate(subUrl)) {
            nodes.push({
              type: HyperNodeType.text,
              text: subUrl,
            });
          } else {
            nodes.push({
              type: HyperNodeType.link,
              url: subUrl,
            });
          }

          cur = result.index + subUrl.length;
          urlRegExp.lastIndex = cur;
        } else {
          const skipStr = content.slice(0, 1);

          nodes.push({
            type: HyperNodeType.text,
            text: skipStr,
          });

          cur = result.index + skipStr.length;
          urlRegExp.lastIndex = cur;
        }
      }
    }
  }
  if (cur !== str.length) {
    nodes.push({ type: HyperNodeType.text, text: str.slice(cur) });
  }
  return nodes;
};
