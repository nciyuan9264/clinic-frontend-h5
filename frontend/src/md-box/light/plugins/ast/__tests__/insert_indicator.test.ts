import { generateAstProcessor } from './utils';
import { InsertIndicatorAstPlugin } from '../insert_indicator';
import { parseMarkdown } from '../../../utils';

describe('测试标记插入函数（insertIndicator）', () => {
  const processor = generateAstProcessor(InsertIndicatorAstPlugin);

  test('在链接后插入标记', () => {
    const ast = parseMarkdown(`
这是[完整的链接](https://bytedance.com)
`);

    expect(processor(ast)).toMatchSnapshot();
  });

  test('在一级列表后插入标记', () => {
    const ast = parseMarkdown(`
- 第一节

- 第二节
`);

    expect(processor(ast)).toMatchSnapshot();
  });

  test('在二级列表后插入标记', () => {
    const ast = parseMarkdown(`
- 第一节

- 第二节

    - 第2.1节

    - 第2.2节
    `);

    expect(processor(ast)).toMatchSnapshot();
  });

  test('在普通段落后插入标记', () => {
    const ast = parseMarkdown(`
测试段落
    `);

    expect(processor(ast)).toMatchSnapshot();
  });

  test('在普通段落内的行内代码块插入标记', () => {
    const ast = parseMarkdown(`
测试段落\`这是一个行内代码块\`
`);

    expect(processor(ast)).toMatchSnapshot();
  });

  test('在代码块后跳过插入标记', () => {
    const ast = parseMarkdown(`
这是第一行

\`\`\`haskell
测试代码
\`\`\`
    `);

    expect(processor(ast)).toMatchSnapshot();
  });

  test('末尾图片后跳过插入标记', () => {
    const ast = parseMarkdown(`
这是第一行

![image](https://pic1.zhimg.com/v2-b444070848d54baf536222b22a51fba4_b.jpg)
    `);

    expect(processor(ast)).toMatchSnapshot();
  });

  test('综合插入场景', () => {
    const ast = parseMarkdown(`
这是[完整的链接](https://bytedance.com)

测试段落\`这是一个行内代码块\`

\`\`\`haskell
测试代码
\`\`\`

- 第一节

- 第二节

    - 第2.1节

    - 第2.2节
`);

    expect(processor(ast)).toMatchSnapshot();
  });
});
