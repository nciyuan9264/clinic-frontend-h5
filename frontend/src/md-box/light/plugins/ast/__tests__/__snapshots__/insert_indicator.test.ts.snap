// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`测试标记插入函数（insertIndicator） 在一级列表后插入标记 1`] = `
{
  "children": [
    {
      "children": [
        {
          "checked": null,
          "children": [
            {
              "children": [
                {
                  "position": {
                    "end": {
                      "column": 6,
                      "line": 2,
                      "offset": 6,
                    },
                    "start": {
                      "column": 3,
                      "line": 2,
                      "offset": 3,
                    },
                  },
                  "type": "text",
                  "value": "第一节",
                },
              ],
              "position": {
                "end": {
                  "column": 6,
                  "line": 2,
                  "offset": 6,
                },
                "start": {
                  "column": 3,
                  "line": 2,
                  "offset": 3,
                },
              },
              "type": "paragraph",
            },
          ],
          "position": {
            "end": {
              "column": 6,
              "line": 2,
              "offset": 6,
            },
            "start": {
              "column": 1,
              "line": 2,
              "offset": 1,
            },
          },
          "spread": false,
          "type": "listItem",
        },
        {
          "checked": null,
          "children": [
            {
              "children": [
                {
                  "position": {
                    "end": {
                      "column": 6,
                      "line": 4,
                      "offset": 13,
                    },
                    "start": {
                      "column": 3,
                      "line": 4,
                      "offset": 10,
                    },
                  },
                  "type": "text",
                  "value": "第二节",
                },
                {
                  "type": "html",
                  "value": "<span class="indicator" />",
                },
              ],
              "position": {
                "end": {
                  "column": 6,
                  "line": 4,
                  "offset": 13,
                },
                "start": {
                  "column": 3,
                  "line": 4,
                  "offset": 10,
                },
              },
              "type": "paragraph",
            },
          ],
          "position": {
            "end": {
              "column": 6,
              "line": 4,
              "offset": 13,
            },
            "start": {
              "column": 1,
              "line": 4,
              "offset": 8,
            },
          },
          "spread": false,
          "type": "listItem",
        },
      ],
      "ordered": false,
      "position": {
        "end": {
          "column": 6,
          "line": 4,
          "offset": 13,
        },
        "start": {
          "column": 1,
          "line": 2,
          "offset": 1,
        },
      },
      "spread": true,
      "start": null,
      "type": "list",
    },
  ],
  "position": {
    "end": {
      "column": 1,
      "line": 5,
      "offset": 14,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`测试标记插入函数（insertIndicator） 在二级列表后插入标记 1`] = `
{
  "children": [
    {
      "children": [
        {
          "checked": null,
          "children": [
            {
              "children": [
                {
                  "position": {
                    "end": {
                      "column": 6,
                      "line": 2,
                      "offset": 6,
                    },
                    "start": {
                      "column": 3,
                      "line": 2,
                      "offset": 3,
                    },
                  },
                  "type": "text",
                  "value": "第一节",
                },
              ],
              "position": {
                "end": {
                  "column": 6,
                  "line": 2,
                  "offset": 6,
                },
                "start": {
                  "column": 3,
                  "line": 2,
                  "offset": 3,
                },
              },
              "type": "paragraph",
            },
          ],
          "position": {
            "end": {
              "column": 6,
              "line": 2,
              "offset": 6,
            },
            "start": {
              "column": 1,
              "line": 2,
              "offset": 1,
            },
          },
          "spread": false,
          "type": "listItem",
        },
        {
          "checked": null,
          "children": [
            {
              "children": [
                {
                  "position": {
                    "end": {
                      "column": 6,
                      "line": 4,
                      "offset": 13,
                    },
                    "start": {
                      "column": 3,
                      "line": 4,
                      "offset": 10,
                    },
                  },
                  "type": "text",
                  "value": "第二节",
                },
              ],
              "position": {
                "end": {
                  "column": 6,
                  "line": 4,
                  "offset": 13,
                },
                "start": {
                  "column": 3,
                  "line": 4,
                  "offset": 10,
                },
              },
              "type": "paragraph",
            },
            {
              "children": [
                {
                  "checked": null,
                  "children": [
                    {
                      "children": [
                        {
                          "position": {
                            "end": {
                              "column": 12,
                              "line": 6,
                              "offset": 26,
                            },
                            "start": {
                              "column": 7,
                              "line": 6,
                              "offset": 21,
                            },
                          },
                          "type": "text",
                          "value": "第2.1节",
                        },
                      ],
                      "position": {
                        "end": {
                          "column": 12,
                          "line": 6,
                          "offset": 26,
                        },
                        "start": {
                          "column": 7,
                          "line": 6,
                          "offset": 21,
                        },
                      },
                      "type": "paragraph",
                    },
                  ],
                  "position": {
                    "end": {
                      "column": 12,
                      "line": 6,
                      "offset": 26,
                    },
                    "start": {
                      "column": 5,
                      "line": 6,
                      "offset": 19,
                    },
                  },
                  "spread": false,
                  "type": "listItem",
                },
                {
                  "checked": null,
                  "children": [
                    {
                      "children": [
                        {
                          "position": {
                            "end": {
                              "column": 12,
                              "line": 8,
                              "offset": 39,
                            },
                            "start": {
                              "column": 7,
                              "line": 8,
                              "offset": 34,
                            },
                          },
                          "type": "text",
                          "value": "第2.2节",
                        },
                        {
                          "type": "html",
                          "value": "<span class="indicator" />",
                        },
                      ],
                      "position": {
                        "end": {
                          "column": 12,
                          "line": 8,
                          "offset": 39,
                        },
                        "start": {
                          "column": 7,
                          "line": 8,
                          "offset": 34,
                        },
                      },
                      "type": "paragraph",
                    },
                  ],
                  "position": {
                    "end": {
                      "column": 12,
                      "line": 8,
                      "offset": 39,
                    },
                    "start": {
                      "column": 5,
                      "line": 8,
                      "offset": 32,
                    },
                  },
                  "spread": false,
                  "type": "listItem",
                },
              ],
              "ordered": false,
              "position": {
                "end": {
                  "column": 5,
                  "line": 9,
                  "offset": 44,
                },
                "start": {
                  "column": 5,
                  "line": 6,
                  "offset": 19,
                },
              },
              "spread": true,
              "start": null,
              "type": "list",
            },
          ],
          "position": {
            "end": {
              "column": 5,
              "line": 9,
              "offset": 44,
            },
            "start": {
              "column": 1,
              "line": 4,
              "offset": 8,
            },
          },
          "spread": true,
          "type": "listItem",
        },
      ],
      "ordered": false,
      "position": {
        "end": {
          "column": 5,
          "line": 9,
          "offset": 44,
        },
        "start": {
          "column": 1,
          "line": 2,
          "offset": 1,
        },
      },
      "spread": true,
      "start": null,
      "type": "list",
    },
  ],
  "position": {
    "end": {
      "column": 5,
      "line": 9,
      "offset": 44,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`测试标记插入函数（insertIndicator） 在代码块后跳过插入标记 1`] = `
{
  "children": [
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 6,
              "line": 2,
              "offset": 6,
            },
            "start": {
              "column": 1,
              "line": 2,
              "offset": 1,
            },
          },
          "type": "text",
          "value": "这是第一行",
        },
      ],
      "position": {
        "end": {
          "column": 6,
          "line": 2,
          "offset": 6,
        },
        "start": {
          "column": 1,
          "line": 2,
          "offset": 1,
        },
      },
      "type": "paragraph",
    },
    {
      "lang": "haskell",
      "meta": null,
      "position": {
        "end": {
          "column": 4,
          "line": 6,
          "offset": 27,
        },
        "start": {
          "column": 1,
          "line": 4,
          "offset": 8,
        },
      },
      "type": "code",
      "value": "测试代码",
    },
  ],
  "position": {
    "end": {
      "column": 5,
      "line": 7,
      "offset": 32,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`测试标记插入函数（insertIndicator） 在普通段落内的行内代码块插入标记 1`] = `
{
  "children": [
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 5,
              "line": 2,
              "offset": 5,
            },
            "start": {
              "column": 1,
              "line": 2,
              "offset": 1,
            },
          },
          "type": "text",
          "value": "测试段落",
        },
        {
          "position": {
            "end": {
              "column": 16,
              "line": 2,
              "offset": 16,
            },
            "start": {
              "column": 5,
              "line": 2,
              "offset": 5,
            },
          },
          "type": "inlineCode",
          "value": "这是一个行内代码块",
        },
        {
          "type": "html",
          "value": "<span class="indicator" />",
        },
      ],
      "position": {
        "end": {
          "column": 16,
          "line": 2,
          "offset": 16,
        },
        "start": {
          "column": 1,
          "line": 2,
          "offset": 1,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 1,
      "line": 3,
      "offset": 17,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`测试标记插入函数（insertIndicator） 在普通段落后插入标记 1`] = `
{
  "children": [
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 5,
              "line": 2,
              "offset": 5,
            },
            "start": {
              "column": 1,
              "line": 2,
              "offset": 1,
            },
          },
          "type": "text",
          "value": "测试段落",
        },
        {
          "type": "html",
          "value": "<span class="indicator" />",
        },
      ],
      "position": {
        "end": {
          "column": 5,
          "line": 2,
          "offset": 5,
        },
        "start": {
          "column": 1,
          "line": 2,
          "offset": 1,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 5,
      "line": 3,
      "offset": 10,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`测试标记插入函数（insertIndicator） 在链接后插入标记 1`] = `
{
  "children": [
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 3,
              "line": 2,
              "offset": 3,
            },
            "start": {
              "column": 1,
              "line": 2,
              "offset": 1,
            },
          },
          "type": "text",
          "value": "这是",
        },
        {
          "children": [
            {
              "position": {
                "end": {
                  "column": 9,
                  "line": 2,
                  "offset": 9,
                },
                "start": {
                  "column": 4,
                  "line": 2,
                  "offset": 4,
                },
              },
              "type": "text",
              "value": "完整的链接",
            },
            {
              "type": "html",
              "value": "<span class="indicator" />",
            },
          ],
          "position": {
            "end": {
              "column": 33,
              "line": 2,
              "offset": 33,
            },
            "start": {
              "column": 3,
              "line": 2,
              "offset": 3,
            },
          },
          "title": null,
          "type": "link",
          "url": "https://bytedance.com",
        },
      ],
      "position": {
        "end": {
          "column": 33,
          "line": 2,
          "offset": 33,
        },
        "start": {
          "column": 1,
          "line": 2,
          "offset": 1,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 1,
      "line": 3,
      "offset": 34,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`测试标记插入函数（insertIndicator） 末尾图片后跳过插入标记 1`] = `
{
  "children": [
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 6,
              "line": 2,
              "offset": 6,
            },
            "start": {
              "column": 1,
              "line": 2,
              "offset": 1,
            },
          },
          "type": "text",
          "value": "这是第一行",
        },
      ],
      "position": {
        "end": {
          "column": 6,
          "line": 2,
          "offset": 6,
        },
        "start": {
          "column": 1,
          "line": 2,
          "offset": 1,
        },
      },
      "type": "paragraph",
    },
    {
      "children": [
        {
          "alt": "image",
          "position": {
            "end": {
              "column": 75,
              "line": 4,
              "offset": 82,
            },
            "start": {
              "column": 1,
              "line": 4,
              "offset": 8,
            },
          },
          "title": null,
          "type": "image",
          "url": "https://pic1.zhimg.com/v2-b444070848d54baf536222b22a51fba4_b.jpg",
        },
      ],
      "position": {
        "end": {
          "column": 75,
          "line": 4,
          "offset": 82,
        },
        "start": {
          "column": 1,
          "line": 4,
          "offset": 8,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 5,
      "line": 5,
      "offset": 87,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`测试标记插入函数（insertIndicator） 综合插入场景 1`] = `
{
  "children": [
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 3,
              "line": 2,
              "offset": 3,
            },
            "start": {
              "column": 1,
              "line": 2,
              "offset": 1,
            },
          },
          "type": "text",
          "value": "这是",
        },
        {
          "children": [
            {
              "position": {
                "end": {
                  "column": 9,
                  "line": 2,
                  "offset": 9,
                },
                "start": {
                  "column": 4,
                  "line": 2,
                  "offset": 4,
                },
              },
              "type": "text",
              "value": "完整的链接",
            },
          ],
          "position": {
            "end": {
              "column": 33,
              "line": 2,
              "offset": 33,
            },
            "start": {
              "column": 3,
              "line": 2,
              "offset": 3,
            },
          },
          "title": null,
          "type": "link",
          "url": "https://bytedance.com",
        },
      ],
      "position": {
        "end": {
          "column": 33,
          "line": 2,
          "offset": 33,
        },
        "start": {
          "column": 1,
          "line": 2,
          "offset": 1,
        },
      },
      "type": "paragraph",
    },
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 5,
              "line": 4,
              "offset": 39,
            },
            "start": {
              "column": 1,
              "line": 4,
              "offset": 35,
            },
          },
          "type": "text",
          "value": "测试段落",
        },
        {
          "position": {
            "end": {
              "column": 16,
              "line": 4,
              "offset": 50,
            },
            "start": {
              "column": 5,
              "line": 4,
              "offset": 39,
            },
          },
          "type": "inlineCode",
          "value": "这是一个行内代码块",
        },
      ],
      "position": {
        "end": {
          "column": 16,
          "line": 4,
          "offset": 50,
        },
        "start": {
          "column": 1,
          "line": 4,
          "offset": 35,
        },
      },
      "type": "paragraph",
    },
    {
      "lang": "haskell",
      "meta": null,
      "position": {
        "end": {
          "column": 4,
          "line": 8,
          "offset": 71,
        },
        "start": {
          "column": 1,
          "line": 6,
          "offset": 52,
        },
      },
      "type": "code",
      "value": "测试代码",
    },
    {
      "children": [
        {
          "checked": null,
          "children": [
            {
              "children": [
                {
                  "position": {
                    "end": {
                      "column": 6,
                      "line": 10,
                      "offset": 78,
                    },
                    "start": {
                      "column": 3,
                      "line": 10,
                      "offset": 75,
                    },
                  },
                  "type": "text",
                  "value": "第一节",
                },
              ],
              "position": {
                "end": {
                  "column": 6,
                  "line": 10,
                  "offset": 78,
                },
                "start": {
                  "column": 3,
                  "line": 10,
                  "offset": 75,
                },
              },
              "type": "paragraph",
            },
          ],
          "position": {
            "end": {
              "column": 6,
              "line": 10,
              "offset": 78,
            },
            "start": {
              "column": 1,
              "line": 10,
              "offset": 73,
            },
          },
          "spread": false,
          "type": "listItem",
        },
        {
          "checked": null,
          "children": [
            {
              "children": [
                {
                  "position": {
                    "end": {
                      "column": 6,
                      "line": 12,
                      "offset": 85,
                    },
                    "start": {
                      "column": 3,
                      "line": 12,
                      "offset": 82,
                    },
                  },
                  "type": "text",
                  "value": "第二节",
                },
              ],
              "position": {
                "end": {
                  "column": 6,
                  "line": 12,
                  "offset": 85,
                },
                "start": {
                  "column": 3,
                  "line": 12,
                  "offset": 82,
                },
              },
              "type": "paragraph",
            },
            {
              "children": [
                {
                  "checked": null,
                  "children": [
                    {
                      "children": [
                        {
                          "position": {
                            "end": {
                              "column": 12,
                              "line": 14,
                              "offset": 98,
                            },
                            "start": {
                              "column": 7,
                              "line": 14,
                              "offset": 93,
                            },
                          },
                          "type": "text",
                          "value": "第2.1节",
                        },
                      ],
                      "position": {
                        "end": {
                          "column": 12,
                          "line": 14,
                          "offset": 98,
                        },
                        "start": {
                          "column": 7,
                          "line": 14,
                          "offset": 93,
                        },
                      },
                      "type": "paragraph",
                    },
                  ],
                  "position": {
                    "end": {
                      "column": 12,
                      "line": 14,
                      "offset": 98,
                    },
                    "start": {
                      "column": 5,
                      "line": 14,
                      "offset": 91,
                    },
                  },
                  "spread": false,
                  "type": "listItem",
                },
                {
                  "checked": null,
                  "children": [
                    {
                      "children": [
                        {
                          "position": {
                            "end": {
                              "column": 12,
                              "line": 16,
                              "offset": 111,
                            },
                            "start": {
                              "column": 7,
                              "line": 16,
                              "offset": 106,
                            },
                          },
                          "type": "text",
                          "value": "第2.2节",
                        },
                        {
                          "type": "html",
                          "value": "<span class="indicator" />",
                        },
                      ],
                      "position": {
                        "end": {
                          "column": 12,
                          "line": 16,
                          "offset": 111,
                        },
                        "start": {
                          "column": 7,
                          "line": 16,
                          "offset": 106,
                        },
                      },
                      "type": "paragraph",
                    },
                  ],
                  "position": {
                    "end": {
                      "column": 12,
                      "line": 16,
                      "offset": 111,
                    },
                    "start": {
                      "column": 5,
                      "line": 16,
                      "offset": 104,
                    },
                  },
                  "spread": false,
                  "type": "listItem",
                },
              ],
              "ordered": false,
              "position": {
                "end": {
                  "column": 12,
                  "line": 16,
                  "offset": 111,
                },
                "start": {
                  "column": 5,
                  "line": 14,
                  "offset": 91,
                },
              },
              "spread": true,
              "start": null,
              "type": "list",
            },
          ],
          "position": {
            "end": {
              "column": 12,
              "line": 16,
              "offset": 111,
            },
            "start": {
              "column": 1,
              "line": 12,
              "offset": 80,
            },
          },
          "spread": true,
          "type": "listItem",
        },
      ],
      "ordered": false,
      "position": {
        "end": {
          "column": 12,
          "line": 16,
          "offset": 111,
        },
        "start": {
          "column": 1,
          "line": 10,
          "offset": 73,
        },
      },
      "spread": true,
      "start": null,
      "type": "list",
    },
  ],
  "position": {
    "end": {
      "column": 1,
      "line": 17,
      "offset": 112,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;
